File: app/src/main/java/com/example/bookworm/fragments/fragment_feed.java
package com.example.bookworm.fragments;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;


import com.example.bookworm.Feed.Comments.DiffUtilCallback;
import com.example.bookworm.Feed.items.Feed;
import com.example.bookworm.Feed.items.FeedAdapter;
import com.example.bookworm.databinding.FragmentFeedBinding;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.bookworm.Feed.subActivity_Feed_Create;
import com.example.bookworm.databinding.LayoutTopbarBinding;
import com.example.bookworm.modules.FBModule;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firestore.v1.Precondition;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class fragment_feed extends Fragment {
    FragmentFeedBinding binding;
    public FeedAdapter feedAdapter;
    private final int LIMIT = 5;
    public ArrayList<Feed> feedList;
    public  static Context mContext;
    private DocumentSnapshot lastVisible; //마지막에 가져온 값 부터 추가로 가져올 수 있도록 함.
    private Map map;
    private FBModule fbModule;
    //isLoading:스크롤을 당겨서 추가로 로딩 중인지 여부를 확인하는 변수
    //canLoad:더 로딩이 가능한지 확인하는 변수[자료의 끝을 판별한다.]
    private Boolean isRefreshing = false, isLoading = false, canLoad = true;
    private int page = 1;
    public ActivityResultLauncher<Intent> startActivityResult = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == subActivity_Feed_Create.CREATE_OK) {
                    pageRefresh();
                }
                if (result.getResultCode() == 26) {
                    ArrayList newList = new ArrayList(feedList);
                    Feed item = (Feed) result.getData().getSerializableExtra("modifiedFeed");
                    newList.remove(item.getPosition());
                    newList.add(item.getPosition(),item);
                    replaceItem(newList);
                }
            });

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mContext = getContext();

        binding = FragmentFeedBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        binding.recyclerView.setNestedScrollingEnabled(false);
        fbModule = new FBModule(getContext());
        fbModule.setLIMIT(LIMIT); //한번에 보여줄 피드의 최대치를 설정
        //Create New Feed
        LayoutTopbarBinding.bind(binding.getRoot()).imgCreatefeed.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(getContext(), subActivity_Feed_Create.class);
                startActivityResult.launch(intent);
            }
        });
        //스와이프하여 새로고침
        binding.swiperefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                pageRefresh();
                isRefreshing = true;
            }
        });
        binding.recyclerView.setItemViewCacheSize(3);
        //피드 초기 호출
        pageRefresh();

        // Inflate the layout for this fragment
        return view;
    }

    //뷰보다 프레그먼트의 생명주기가 길어서, 메모리 누수 발생가능
    //누수 방지를 위해 뷰가 Destroy될 때, binding값을 nullify 함.
    @Override
    public void onDestroyView() {
        binding = null;
        super.onDestroyView();
    }

    //리사이클러뷰를 초기화
    private void initRecyclerView() {
        binding.recyclerView.setAdapter(feedAdapter);
        initScrollListener(); //무한스크롤
    }

    private void initAdapter() {
        feedAdapter = new FeedAdapter(feedList, getContext());
    }

    //리사이클러뷰 스크롤 초기화
    private void initScrollListener() {
        binding.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                int lastVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition();
                if (!isLoading) {
                    try {
                        if (layoutManager != null && lastVisibleItemPosition == feedAdapter.getItemCount() - 1) {
                            deleteLoading();
                            //이전에 가져왔던 자료를 인자로 보내주어 그 다음 자료부터 조회한다.
                            map.put("lastVisible", lastVisible);
                            //쿼리를 보내어, 데이터를 조회한다.
                            fbModule.readData(1, map, null);
                            //현재 로딩을 끝냄을 알린다.
                            isLoading = true;
                        }
                    } catch (NullPointerException e) {

                    }
                }

            }
        });

    }

    //페이지 새로고침 시 사용하는 메소드
    private void pageRefresh() {
        initFeed();
        map = new HashMap();
//        String[] followers={"113371714153801966916","2091385654"};
//        map.put("followers",Arrays.asList(followers));
        if (map.get("lastVisible") != null) map.remove("lastVisible");
        feedList = new ArrayList<>(); //챌린지를 담는 리스트 생성
        isLoading = true;
        fbModule.readData(1, map, null);

    }

    //피드 초기화
    private void initFeed() {
        isLoading = true;
        page = 1;
        canLoad = true;
        lastVisible = null;
        initRecyclerView();
    }

    // 로딩이 완료되면 프로그레스바를 지움
    public void deleteLoading() {
        ArrayList arr = new ArrayList(feedList);
        arr.remove(arr.size() - 1);
        replaceItem(arr); //데이터가 삭제됨을 알림.
    }

    public void replaceItem(ArrayList newthings) {
        DiffUtilCallback callback = new DiffUtilCallback(feedList, newthings);
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(callback, true);
        feedList.clear();
        feedList.addAll(newthings);
        feedAdapter.setData(feedList);
        diffResult.dispatchUpdatesTo(feedAdapter);
    }

    //fb모듈을 통해 전달받은 값을 세팅
    public void moduleUpdated(List<DocumentSnapshot> a, ArrayList<DocumentSnapshot> b) {
        ArrayList newList = new ArrayList(feedList); //새로 들어오는 데이터와 기존 데이터의 길이를 비교하기 위해서
        if (isRefreshing) {
            binding.swiperefresh.setRefreshing(false);
            isRefreshing = false;
        }
        if (page == 1) {
            isLoading = false;
            feedList = new ArrayList<>(); //챌린지를 담는 리스트 생성
        }
        if (a == null) {
            if (page == 1) {
                Toast.makeText(getContext(), "화면에 표시할 포스트가 없습니다.", Toast.LENGTH_SHORT).show();
                feedList = new ArrayList<>(); //챌린지를 담는 리스트 생성
                initFeed();
                initAdapter();
            } else {
                canLoad = false;
                isLoading = true;
            }

        } else {
            //가져온 데이터를 for문을 이용하여, feed리스트에 차곡차곡 담는다.
            try {
                for (int i = 0; i < a.size(); i++) {
                    Map Adata = a.get(i).getData();//피드 데이터
                    Map Bdata = b.get(i) == null ? null : b.get(i).getData(); //상단의 댓글 데이터
                    Feed feed = new Feed();
                    feed.setData(Adata, Bdata);
                    newList.add(feed);
                }
                //가져온 값의 마지막 snapshot부터 이어서 가져올 수 있도록 하기 위함.
                lastVisible = a.get(a.size() - 1);
                //리사이클러뷰에서 튕김현상이 발생하여 넣어준 코드
                //현재 불러오는 값의 크기(a.size())가 페이징 제한 값(LIMIT)보다 작은 경우 => 더이상 불러오지 않게 함.
                if (a.size() < LIMIT) {
                    canLoad = false;
                }
            } catch (NullPointerException e) {
                canLoad = false;
            }
        }
        //만약 더이상 불러오지 못 할 경우
        if (canLoad == false) {
            isLoading = true;
            if (page > 1) replaceItem(newList);//이미 불러온 데이터가 있는 경우엔 가져온 데이터 만큼의 범위를 늘려준다.
            else { //없는 경우엔 새로운 어댑터에 데이터를 담아서 띄워준다.
                initAdapter(); //어댑터 초기화
                replaceItem(newList);
                initRecyclerView(); //리사이클러뷰에 띄워주기
            }
        }
        //더 불러올 데이터가 있는 경우
        else {
            isLoading = false;
            newList.add(new Feed());//로딩바 표시를 위한 빈 값
            if (page > 1) replaceItem(newList);
            else {
                initAdapter();//어댑터 초기화
                replaceItem(newList);
                initRecyclerView(); //리사이클러뷰에 띄워주기
            }
            page++; //로딩을 다하면 그다음 페이지로 넘어간다.
        }
    }
}

File: app/src/main/java/com/example/bookworm/bottomMenu/feed/FragmentFeed.kt
package com.example.bookworm.bottomMenu.feed

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.lifecycle.*
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.bookworm.DdayCounter
import com.example.bookworm.LoadState
import com.example.bookworm.R
import com.example.bookworm.appLaunch.views.MainActivity
import com.example.bookworm.bottomMenu.feed.comments.SubActivityComment
import com.example.bookworm.bottomMenu.profile.UserInfoViewModel
import com.example.bookworm.core.dataprocessing.image.ImageProcessing
import com.example.bookworm.databinding.FragmentFeedTopbarBinding
import com.example.bookworm.databinding.TmpActivityFeedBinding
import kotlinx.coroutines.launch
import androidx.recyclerview.widget.RecyclerView.OnScrollListener as OnScrollListener


class FragmentFeed : Fragment() {
    private lateinit var binding: TmpActivityFeedBinding //Binding

    private val feedViewModel by lazy {
        ViewModelProvider(this, FeedViewModel.Factory(requireContext()))[FeedViewModel::class.java]
    }
    val feedAdapter by lazy {
        FeedAdapter()
    }
    private var isDataEnd = false // 파이어베이스 내에서 검색할 때, 데이터 끝인지 판별하는 변수
    private var need2Mov = -1
    private var storiesBar: RecyclerView? = null


    //다른 액티비티로부터의 결과값을 받기 위함
    var startActivityResult = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == SubActivityCreatePost.CREATE_OK) {
            val item = result.data!!.getParcelableExtra<Feed>("feedData")
            need2Mov = 0 //상단으로 스크롤 하기 위한 플래그 설정
            feedAdapter.currentList.toList().apply {
                feedAdapter.submitList(listOf(item) + this)
            }
            Toast.makeText(context,
                    "게시물이 업로드 되었습니다.", Toast.LENGTH_SHORT)
                    .show()

        }
        /**
         * 게시물이 수정된 경우
         * */
        if (result.resultCode == SubActivityModifyPost.MODIFY_OK || result.resultCode == SubActivityComment.FEED_MODIFIED) {
            //수정된 아이템
            result.data!!.apply {
                //게시물 업로드 하는 함수
                val item = getParcelableExtra<Feed>("modifiedFeed")!!
                feedViewModel.uploadFeed(item, null, ImageProcessing(requireContext()), item.imgurl)
                feedViewModel.nowFeedUploadState.observe(viewLifecycleOwner, Observer {
                    if (it == LoadState.Done) {
                        //수정 업데이트 적용
                        need2Mov = item.position
                        item.duration = DdayCounter.getDuration(item.date!!)
                        feedAdapter.currentList.toMutableList().apply {
                            this.removeAt(need2Mov)
                            add(need2Mov, item)
                            feedAdapter.submitList(this.toList())
                        }
                        Toast.makeText(requireContext(), "게시물이 정상적으로 수정되었습니다. ", Toast.LENGTH_SHORT).show()
                    }
                })

            }
        }
        //댓글 화면에서 게시물을 삭제하려고 한 경우
        if (result.resultCode == SubActivityComment.FEED_DELETE) {
            result.data!!.apply {
                val target = getParcelableExtra<Feed>("deleteTarget")!!
                feedAdapter.currentList.toMutableList().apply {
                    this.removeAt(target.position)
                    feedAdapter.submitList(this.toList())
                }
                Toast.makeText(context, "게시물이 정상적으로 삭제되었습니다.", Toast.LENGTH_SHORT).show()
            }
        }

    }

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?,
    ): View {
        binding = TmpActivityFeedBinding.inflate(layoutInflater)
        getFeeds(true) // 서버로부터 피드 정보를 받아옴

        //피드 새로 만들기
        viewLifecycleOwner.lifecycle.coroutineScope.launch {
            val data = ViewModelProvider(context as MainActivity, UserInfoViewModel.Factory(requireContext()))[UserInfoViewModel::class.java]
            val mainUser = data.suspendGetUser(null)
            FragmentFeedTopbarBinding.bind(binding.root).apply {
                imgCreatefeed.setOnClickListener {
                    val intent = Intent(context, SubActivityCreatePost::class.java)
                    intent.putExtra("mainUser", mainUser)
                    startActivityResult.launch(intent)
                }
                tvLogo.setOnClickListener {
                    binding.recyclerView.smoothScrollToPosition(0)
                }
            }

        }

        setAdapter()


        return binding.root
    }

    private fun check2Top() {
        //게시물 작성 완료 시 or 수정 완료시
        if (need2Mov >= 0) {
            binding.recyclerView.scrollToPosition(need2Mov)
            need2Mov = -1 //기본값으로 초기화
        }
    }

    private fun processFeedDelete(pos: Int) {
        val mutableList = feedAdapter.currentList.toMutableList()
        mutableList.removeAt(pos)
        if (!isDataEnd) {
            val liveData = MutableLiveData<Feed>()
            feedViewModel.loadPost(liveData)
            liveData.observe(viewLifecycleOwner, Observer { addedItem ->
                mutableList.removeLast()
                mutableList += listOf(addedItem, Feed())
                feedAdapter.submitList(mutableList)
            })
        } else
            feedAdapter.submitList(mutableList)

    }

    private fun setAdapter() {
        binding.apply {
            recyclerView.adapter = feedAdapter //어댑터 세팅
            //데이터 변경시 감지 후 자동 스크롤
            feedAdapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
                override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                    check2Top()
                }

                override fun onChanged() {
                    check2Top()
                }
            })
            feedAdapter.setFeedMenuListener(object : OnFeedMenuClickListener {
                override fun onItemClick(holder: FeedViewHolder, view: View, position: Int) {
                    val feed = feedAdapter.currentList[position]
                    feed.position = position
                    val popupMenu = customMenuPopup(context!!, view)
                    if (feed.isUserPost) {
                        popupMenu.setItem(feed)
                        popupMenu.liveState.observe(viewLifecycleOwner, Observer { data ->
                            if (data == popupMenu.FEED_DELETE) {
                                //게시물 삭제시 새로운 게시물 하나를 더 불러옴.
                                processFeedDelete(position)
                            }
                        })
                    }
                }
            })

//            recyclerView.itemAnimator = null //리사이클러뷰 애니메이션 제거
            recyclerView.isNestedScrollingEnabled = false
            recyclerView.setHasFixedSize(true)



            swiperefresh.setOnRefreshListener {
                pageRefresh()
            } //스와이프하여 새로고침
            with(recyclerView) {
                addOnScrollListener(object : OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        val layoutManager = recyclerView.layoutManager as LinearLayoutManager?
                        val lastVisibleItemPosition =
                                layoutManager!!.findLastCompletelyVisibleItemPosition()
                        if ((lastVisibleItemPosition
                                        == feedAdapter.currentList.lastIndex) && lastVisibleItemPosition > 0)
                            getFeeds(false)
                    }
                })
            }
        }
    }

    //화면 이동 시 보여졌던 키보드를 숨김
    override fun onHiddenChanged(hidden: Boolean) {
        val imm =
                requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (binding.root.findViewById<View?>(R.id.edtComment) != null) imm.hideSoftInputFromWindow(
                binding.root.findViewById<View>(R.id.edtComment).windowToken,
                InputMethodManager.HIDE_NOT_ALWAYS
        )
        super.onHiddenChanged(hidden)
    }


    //페이지 새로고침 시 사용하는 메소드
    fun pageRefresh() {
        showShimmer(true)
        isDataEnd = false
        getFeeds(true)
        feedAdapter.submitList(emptyList())
    }


    //FireStore에서 피드 가져옴
    private fun getFeeds(refreshing: Boolean) {
        if (!isDataEnd) {
            feedViewModel.loadPosts(refreshing)
            feedViewModel.nowFeedLoadState.observe(viewLifecycleOwner, Observer { nowState ->
                //피드를 새로 불러올 때 활성화
                showShimmer(
                        if (!refreshing) false
                        else nowState == LoadState.Loading)
                //데이터 로딩이 다 되었다면
                if (nowState == LoadState.Done) {
                    var current = feedAdapter.currentList.toMutableList() //기존에 가지고 있던 아이템 목록
                    if (refreshing) {
                        current.clear()
                        binding.swiperefresh.isRefreshing = false //새로고침인 경우, 데이터가 다 로딩 된 후 새로고침 표시 없애기
                    }
                    //만약 현재 목록이 비어있지 않고, 마지막 아이템이 로딩 아이템 이라면 마지막 아이템을 제거
                    if (current.isNotEmpty() && current.last().feedID == null) current.removeLast()
                    //데이터의 끝에 다다르지 않았다면, 현재 목록에 불러온 아이템을 추가한다.
                    if (feedViewModel.postsData != null && !current.containsAll(feedViewModel.postsData!!)) {
                        val resultData = feedViewModel.postsData!!.toMutableList()
                        val i = resultData.iterator()
                        //반복하면서 중복되는 데이터가 있는 경우 삭제
                        while (i.hasNext()) {
                            if (current.contains(i.next())) i.remove()
                        }
                        current.addAll(resultData)
                        current.add(Feed())
                    }
                    //데이터의 끝에 다다랐다면 끝이라는 것을 변수에 저장
                    else isDataEnd = true

                    //변경된 리스트를 어댑터에 반영
                    feedAdapter.submitList(current)
                }
            })
        }
    }


    //shimmer을 켜고 끄고 하는 메소드
    fun showShimmer(bool: Boolean) {
        binding.llFeed.isVisible = !bool
        if (bool) binding.SFLFeed.startShimmer() else binding.SFLFeed.stopShimmer()
        binding.SFLFeed.isVisible = bool
    }
}